File: typescript/src/hosting/neuron/router.ts
// hosting/neuron/router.ts
//
// The concierge router — TS port of CircleAI.Hosting.Neuron's INeuronRouter +
// HeuristicNeuronRouter + NeuronGate. Per turn it decides, with cheap
// keyword/length heuristics (never a model), whether the always-warm generalist
// answers or a capability-matched specialist should be hot-loaded. This is the
// "gear system": load and run only what the turn needs.

import { ChatCapability } from "../../inference/index.js";

/** Which resident organ should serve a turn. Mirrors the Organ enum. */
export enum Organ {
  Generalist = "generalist",
  Specialist = "specialist",
}

/** What the router sees for one turn. Mirrors RouteContext. */
export interface RouteContext {
  /** The user's text for this turn. */
  readonly query: string;
  /** True when the turn carries image bytes (vision). */
  readonly hasImage?: boolean;
  /** Total prompt length in chars; defaults to query.length. */
  readonly promptChars?: number;
}

/** The router's per-turn verdict. Mirrors RouteDecision. */
export interface RouteDecision {
  readonly organ: Organ;
  /** Capability the specialist must satisfy (ignored for the generalist). */
  readonly capability: ChatCapability | number;
  /** Short human-readable reason, for observability. */
  readonly reason: string;
}

/** The plain-generalist decision. */
export function generalistDecision(reason = "generalist"): RouteDecision {
  return { organ: Organ.Generalist, capability: ChatCapability.Default, reason };
}

/** A specialist decision for a capability. */
export function specialistDecision(
  capability: ChatCapability | number,
  reason: string,
): RouteDecision {
  return { organ: Organ.Specialist, capability, reason };
}

/** Per-turn router contract. Mirrors INeuronRouter. */
export interface INeuronRouter {
  route(context: RouteContext): RouteDecision;
}

/**
 * Safety/capability veto over a specialist pick. Mirrors NeuronGate: when the
 * predicate rejects a decision the router demotes it to the generalist, so a
 * turn is never blocked — the floor always answers.
 */
export class NeuronGate {
  constructor(
    private readonly allowSpecialist: (decision: RouteDecision) => boolean = () =>
      true,
  ) {}

  /** True if the specialist decision is permitted. */
  allows(decision: RouteDecision): boolean {
    return this.allowSpecialist(decision);
  }
}

/** Reasoning cue substrings → route to a Reasoning specialist. */
const REASONING_CUES: readonly string[] = [
  "debug",
  "stack trace",
  "solve",
  "prove",
  "reason",
  "analy",
  "calculate",
  "equation",
  "step by step",
  "algorithm",
  "why does",
  "derive",
  "diagnose",
];

/**
 * The default concierge router. Cheap heuristics, in priority order:
 *   image bytes     → Specialist(Vision)
 *   long prompt     → Specialist(LongContext)   (≥ longContextChars)
 *   reasoning cues  → Specialist(Reasoning)
 *   otherwise       → Generalist(Default)
 * A NeuronGate veto demotes any specialist pick back to the generalist.
 * Mirrors HeuristicNeuronRouter.
 */
export class HeuristicNeuronRouter implements INeuronRouter {
  private readonly longContextChars: number;
  private readonly gate: NeuronGate;

  constructor(opts: { longContextChars?: number; gate?: NeuronGate } = {}) {
    this.longContextChars = opts.longContextChars ?? 4000;
    this.gate = opts.gate ?? new NeuronGate();
  }

  route(context: RouteContext): RouteDecision {
    const decision = this.classify(context);
    if (decision.organ === Organ.Specialist && !this.gate.allows(decision)) {
      return generalistDecision("gate-vetoed → generalist");
    }
    return decision;
  }

  private classify(context: RouteContext): RouteDecision {
    if (context.hasImage) {
      return specialistDecision(ChatCapability.Vision, "image present");
    }
    const chars = context.promptChars ?? context.query.length;
    if (chars >= this.longContextChars) {
      return specialistDecision(
        ChatCapability.LongContext,
        `long prompt (${chars} chars)`,
      );
    }
    const q = context.query.toLowerCase();
    if (REASONING_CUES.some((c) => q.includes(c))) {
      return specialistDecision(ChatCapability.Reasoning, "reasoning cue");
    }
    return generalistDecision("default");
  }
}

File: python/src/circle_ai/hosting/neuron/router.py
"""Concierge router — port of CircleAI.Hosting.Neuron router + gate.

Per turn, decide whether the always-warm generalist answers or a
capability-matched specialist should. Cheap heuristics, no model inference.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
from typing import Callable, Optional, Protocol, runtime_checkable

from ...inference.inference import ChatCapability

__all__ = [
    "Organ",
    "RouteContext",
    "RouteDecision",
    "INeuronRouter",
    "NeuronGate",
    "HeuristicNeuronRouter",
]


class Organ(IntEnum):
    """Which organ answers a turn."""

    GENERALIST = 0
    SPECIALIST = 1


@dataclass(frozen=True)
class RouteContext:
    """Inputs the concierge classifies for a single turn."""

    query: str
    has_image: bool = False


@dataclass(frozen=True)
class RouteDecision:
    """The concierge's per-turn decision."""

    organ: Organ
    capability: ChatCapability
    reason: str

    @classmethod
    def generalist(cls, reason: str = "generalist") -> "RouteDecision":
        return cls(Organ.GENERALIST, ChatCapability.DEFAULT, reason)

    @classmethod
    def specialist(cls, capability: ChatCapability, reason: str) -> "RouteDecision":
        return cls(Organ.SPECIALIST, capability, reason)


@runtime_checkable
class INeuronRouter(Protocol):
    """The concierge's decision layer. Mirrors ``INeuronRouter``."""

    def route(self, context: RouteContext) -> RouteDecision: ...


class NeuronGate:
    """Guardrail checkpoint. An optional predicate can force a turn back to the
    generalist. ``None`` applies no veto — the honest default.
    """

    def __init__(self, allow_specialist: Optional[Callable[[str], bool]] = None) -> None:
        self._allow = allow_specialist

    def apply(self, decision: RouteDecision, context: RouteContext) -> RouteDecision:
        if (
            decision.organ == Organ.SPECIALIST
            and self._allow is not None
            and not self._allow(context.query)
        ):
            return RouteDecision.generalist("gate: specialist vetoed -> generalist")
        return decision


# Lowercase substrings that signal a turn wants an explicit reasoning model.
_REASONING_CUES = (
    "prove", "solve", "calculate", "derive", "algorithm", "complexity",
    "debug", "stack trace", "refactor", "regex", "step by step",
    "step-by-step", "theorem", "equation", "big-o", "big o",
)


class HeuristicNeuronRouter:
    """Default router: modality (image -> vision), length (long -> long-context),
    and reasoning cues (-> reasoning); everything else stays on the generalist.
    Mirrors ``HeuristicNeuronRouter``.
    """

    def __init__(
        self, gate: Optional[NeuronGate] = None, long_context_chars: int = 4000
    ) -> None:
        self._gate = gate or NeuronGate()
        self._long_context_chars = long_context_chars if long_context_chars > 0 else 4000

    def route(self, context: RouteContext) -> RouteDecision:
        if context is None:
            raise ValueError("context is required")
        return self._gate.apply(self._classify(context), context)

    def _classify(self, context: RouteContext) -> RouteDecision:
        # 1. An image attachment needs a vision model.
        if context.has_image:
            return RouteDecision.specialist(
                ChatCapability.VISION, "image attached -> vision specialist"
            )

        query = context.query or ""

        # 2. A very long prompt needs a long-context model.
        if len(query) >= self._long_context_chars:
            return RouteDecision.specialist(
                ChatCapability.LONG_CONTEXT,
                f"prompt length {len(query)} >= {self._long_context_chars} "
                f"-> long-context specialist",
            )

        # 3. Reasoning / coding cues want an explicit reasoning model.
        lower = query.lower()
        for cue in _REASONING_CUES:
            if cue in lower:
                return RouteDecision.specialist(
                    ChatCapability.REASONING, f"reasoning cue '{cue}' -> reasoning specialist"
                )

        # 4. Everything else: the always-warm generalist.
        return RouteDecision.generalist("no specialist cue -> generalist")
